File: game-engine/src/test/java/org/luxons/sevenwonders/game/effects/DiscountTest.java
package org.luxons.sevenwonders.game.effects;

import org.junit.Assume;
import org.junit.experimental.theories.DataPoints;
import org.junit.experimental.theories.Theories;
import org.junit.experimental.theories.Theory;
import org.junit.runner.RunWith;
import org.luxons.sevenwonders.game.boards.Board;
import org.luxons.sevenwonders.game.resources.Provider;
import org.luxons.sevenwonders.game.resources.ResourceTransactions;
import org.luxons.sevenwonders.game.resources.ResourceType;
import org.luxons.sevenwonders.game.test.TestUtilsKt;

import static org.junit.Assert.assertEquals;

@RunWith(Theories.class)
public class DiscountTest {

    @DataPoints
    public static int[] discountedPrices() {
        return new int[] {0, 1, 2};
    }

    @DataPoints
    public static ResourceType[] resourceTypes() {
        return ResourceType.values();
    }

    @DataPoints
    public static Provider[] providers() {
        return Provider.values();
    }

    @Theory
    public void apply_givesDiscountedPrice(int discountedPrice, ResourceType discountedType, Provider provider) {
        Board board = TestUtilsKt.testBoard(ResourceType.CLAY, 3);
        Discount discount = new Discount();
        discount.setDiscountedPrice(discountedPrice);
        discount.getProviders().add(provider);
        discount.getResourceTypes().add(discountedType);
        discount.apply(board);

        ResourceTransactions transactions = TestUtilsKt.createTransactions(provider, discountedType);
        assertEquals(discountedPrice, board.getTradingRules().computeCost(transactions));
    }

    @Theory
    public void apply_doesNotAffectOtherResources(int discountedPrice, ResourceType discountedType, Provider provider,
            ResourceType otherType, Provider otherProvider) {
        Assume.assumeTrue(otherProvider != provider);
        Assume.assumeTrue(otherType != discountedType);

        Board board = TestUtilsKt.testBoard(ResourceType.CLAY, 3);
        Discount discount = new Discount();
        discount.setDiscountedPrice(discountedPrice);
        discount.getProviders().add(provider);
        discount.getResourceTypes().add(discountedType);
        discount.apply(board);

        // this is the default in the settings used by TestUtilsKt.testBoard()
        int normalPrice = 2;

        ResourceTransactions fromOtherType = TestUtilsKt.createTransactions(provider, otherType);
        assertEquals(normalPrice, board.getTradingRules().computeCost(fromOtherType));

        ResourceTransactions fromOtherProvider = TestUtilsKt.createTransactions(otherProvider, discountedType);
        assertEquals(normalPrice, board.getTradingRules().computeCost(fromOtherProvider));

        ResourceTransactions fromOtherProviderAndType = TestUtilsKt.createTransactions(otherProvider, otherType);
        assertEquals(normalPrice, board.getTradingRules().computeCost(fromOtherProviderAndType));
    }
}

File: game-engine/src/test/kotlin/org/luxons/sevenwonders/game/effects/DiscountTest.kt
package org.luxons.sevenwonders.game.effects

import org.junit.Assert.assertEquals
import org.junit.Assume
import org.junit.experimental.theories.DataPoints
import org.junit.experimental.theories.Theories
import org.junit.experimental.theories.Theory
import org.junit.runner.RunWith
import org.luxons.sevenwonders.game.resources.Provider
import org.luxons.sevenwonders.game.resources.ResourceType
import org.luxons.sevenwonders.game.test.createTransactions
import org.luxons.sevenwonders.game.test.testBoard

@RunWith(Theories::class)
class DiscountTest {

    @Theory
    fun apply_givesDiscountedPrice(discountedPrice: Int, discountedType: ResourceType, provider: Provider) {
        val board = testBoard(ResourceType.CLAY, 3)
        val discount = Discount(listOf(discountedType), listOf(provider), discountedPrice)
        discount.applyTo(board)

        val transactions = createTransactions(provider, discountedType)
        assertEquals(discountedPrice.toLong(), board.tradingRules.computeCost(transactions).toLong())
    }

    @Theory
    fun apply_doesNotAffectOtherResources(
        discountedPrice: Int, discountedType: ResourceType, provider: Provider,
        otherType: ResourceType, otherProvider: Provider
    ) {
        Assume.assumeTrue(otherProvider != provider)
        Assume.assumeTrue(otherType != discountedType)

        val board = testBoard(ResourceType.CLAY, 3)
        val discount = Discount(listOf(discountedType), listOf(provider), discountedPrice)
        discount.applyTo(board)

        // this is the default in the settings used by TestUtilsKt.testBoard()
        val normalPrice = 2

        val fromOtherType = createTransactions(provider, otherType)
        assertEquals(normalPrice.toLong(), board.tradingRules.computeCost(fromOtherType).toLong())

        val fromOtherProvider = createTransactions(otherProvider, discountedType)
        assertEquals(normalPrice.toLong(), board.tradingRules.computeCost(fromOtherProvider).toLong())

        val fromOtherProviderAndType = createTransactions(otherProvider, otherType)
        assertEquals(normalPrice.toLong(), board.tradingRules.computeCost(fromOtherProviderAndType).toLong())
    }

    companion object {

        @JvmStatic
        @DataPoints
        fun discountedPrices(): IntArray = intArrayOf(0, 1, 2)

        @JvmStatic
        @DataPoints
        fun resourceTypes(): Array<ResourceType> = ResourceType.values()

        @JvmStatic
        @DataPoints
        fun providers(): Array<Provider> = Provider.values()
    }
}
